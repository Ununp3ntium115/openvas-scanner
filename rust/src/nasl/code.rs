use std::path::{Path, PathBuf};

use codespan_reporting::files::SimpleFile;

use super::{
    Loader,
    syntax::{
        Ast, LoadError, Tokenizer,
        parser::{ParseError, Parser},
    },
};

fn parse(code: &str) -> Result<Ast, Vec<ParseError>> {
    let tokenizer = Tokenizer::tokenize(code);
    let mut parser = Parser::new(tokenizer);
    parser.parse_program()
}

pub type SourceFile = SimpleFile<String, String>;

pub struct ParseResult {
    result: Result<Ast, Vec<ParseError>>,
    file: SourceFile,
}

impl ParseResult {
    pub fn new(code: &str, path: &Path) -> Self {
        let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
        let result = parse(code);
        Self { file, result }
    }

    pub fn new_without_file(code: &str) -> Self {
        Self::new(code, Path::new(""))
    }

    pub fn result(self) -> Result<Ast, Vec<ParseError>> {
        self.result
    }

    pub fn num_errors(&self) -> usize {
        match &self.result {
            Ok(_) => 0,
            Err(errs) => errs.len(),
        }
    }

    pub fn emit_errors(self) -> Option<Ast> {
        match self.result {
            Ok(result) => Some(result),
            Err(errors) => {
                super::error::emit_errors(&self.file, errors.into_iter());
                None
            }
        }
    }

    #[cfg(test)]
    pub fn unwrap_errors_str(self) -> String {
        let errs = self.result.unwrap_err();
        super::error::emit_errors_str(&self.file, errs.into_iter())
    }

    pub fn file(&self) -> &SourceFile {
        &self.file
    }
}

pub struct Code {
    code: String,
    path: Option<PathBuf>,
}

impl Code {
    pub fn load(loader: &dyn Loader, path: impl AsRef<Path>) -> Result<Self, LoadError> {
        Ok(Self {
            code: loader.load(&path.as_ref().to_string_lossy())?,
            path: Some(path.as_ref().to_owned()),
        })
    }

    pub fn from_string(code: &str) -> Self {
        Self {
            path: None,
            code: code.to_string(),
        }
    }

    pub fn from_string_fake_filename(code: &str, filename: impl AsRef<Path>) -> Self {
        Self {
            code: code.to_string(),
            path: Some(filename.as_ref().to_owned()),
        }
    }

    #[cfg(test)]
    pub fn tokenize(self) -> TokenizeResult {
        match self.path {
            Some(path) => TokenizeResult::new(&self.code, &path),
            None => unimplemented!(),
        }
    }

    pub fn parse(self) -> ParseResult {
        match self.path {
            Some(path) => ParseResult::new(&self.code, &path),
            None => ParseResult::new_without_file(&self.code),
        }
    }

    pub fn code(&self) -> &str {
        &self.code
    }
}

#[cfg(test)]
pub use tokenize::TokenizeResult;

#[cfg(test)]
mod tokenize {
    use std::path::Path;

    use codespan_reporting::files::SimpleFile;
    use itertools::{Either, Itertools};

    use crate::nasl::syntax::{Token, Tokenizer, TokenizerError};

    use super::{super::error, SourceFile};

    fn split_tokens_and_errors(tokenizer: Tokenizer) -> Result<Vec<Token>, Vec<TokenizerError>> {
        let (tokens, errors): (Vec<_>, Vec<_>) = tokenizer.partition_map(|a| match a {
            Ok(a) => Either::Left(a),
            Err(a) => Either::Right(a),
        });
        if !errors.is_empty() {
            Err(errors)
        } else {
            Ok(tokens)
        }
    }

    pub struct TokenizeResult {
        result: Result<Vec<Token>, Vec<TokenizerError>>,
        file: SourceFile,
    }

    impl TokenizeResult {
        pub fn new(code: &str, path: &Path) -> Self {
            let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
            let result = split_tokens_and_errors(Tokenizer::tokenize(code));
            Self { file, result }
        }

        pub fn result(self) -> Result<Vec<Token>, Vec<TokenizerError>> {
            self.result
        }

        pub fn emit_errors(self) -> Option<Vec<Token>> {
            match self.result {
                Ok(result) => Some(result),
                Err(errors) => {
                    error::emit_errors(&self.file, errors.into_iter());
                    None
                }
            }
        }

        pub fn unwrap_errors_str(self) -> String {
            let errs = self.result.unwrap_err();
            error::emit_errors_str(&self.file, errs.into_iter())
        }
    }
}
