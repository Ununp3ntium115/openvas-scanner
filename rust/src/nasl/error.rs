use std::ops::Range;

use codespan_reporting::diagnostic::Diagnostic;
use codespan_reporting::diagnostic::Label;
use codespan_reporting::term;
use codespan_reporting::term::termcolor::ColorChoice;
use codespan_reporting::term::termcolor::StandardStream;

use super::code::SourceFile;

pub trait AsCodespanError {
    fn range(&self) -> Range<usize>;
    fn message(&self) -> String;
}

pub fn emit_errors<T: AsCodespanError>(file: &SourceFile, errs: impl Iterator<Item = T>) {
    let writer = StandardStream::stderr(ColorChoice::Always);
    let config = codespan_reporting::term::Config::default();
    for err in errs {
        let diagnostic = Diagnostic::error()
            .with_message(&err.message())
            .with_labels(vec![
                Label::primary((), err.range()).with_message(&err.message())
            ]);
        term::emit(&mut writer.lock(), &config, file, &diagnostic).unwrap();
    }
}
