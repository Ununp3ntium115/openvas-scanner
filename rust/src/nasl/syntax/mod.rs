// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
mod cursor;
mod error;
mod grouping_extension;
mod keyword_extension;
mod lexer;
mod loader;
mod naslvalue;
mod operation;
mod prefix_extension;
mod statement;
mod token;
mod tokenizer;
mod variable_extension;

pub use crate::storage::item::ACT;
pub use error::{ErrorKind, SyntaxError};
pub use lexer::Lexer;
pub use loader::*;
pub use naslvalue::*;
pub use statement::*;
pub use token::Keyword;
pub use token::Token;
pub use token::TokenKind;
pub use tokenizer::Tokenizer;

/// Parses given code and returns found Statements and Errors
///
/// # Examples
/// Basic usage:
///
/// ```
/// use scannerlib::nasl::syntax::{Statement, SyntaxError, parse};
/// let statements =
///     parse("a = 23;b = 1;").collect::<Vec<Result<Statement, SyntaxError>>>();
/// ````
pub fn parse(code: &str) -> impl Iterator<Item = Result<Statement, SyntaxError>> + '_ {
    // TODO Do not unwrap here, handle errors properly.
    let tokens = Tokenizer::tokenize(code).unwrap();
    Lexer::new(tokens)
}

#[cfg(test)]
mod tests {
    use crate::nasl::syntax::{
        cursor::Cursor,
        token::{Keyword, Token, TokenKind},
        Tokenizer,
    };

    #[test]
    fn use_cursor() {
        let mut cursor = Cursor::new("  \n\tdisplay(12);");
        cursor.skip_while(|c| c.is_whitespace());
        assert_eq!(cursor.advance(), Some('d'));
    }

    #[test]
    fn use_tokenizer() {
        let all_tokens = Tokenizer::tokenize("local_var hello = 'World!';").unwrap();
        assert_eq!(
            all_tokens,
            vec![
                Token {
                    kind: TokenKind::Identifier(Keyword::LocalVar),
                    line_column: (1, 1),
                    position: (0, 9)
                },
                Token {
                    kind: TokenKind::Identifier(Keyword::Undefined("hello".to_owned())),
                    line_column: (1, 11),
                    position: (10, 15)
                },
                Token {
                    kind: TokenKind::Equal,
                    line_column: (1, 17),
                    position: (16, 17)
                },
                Token {
                    kind: TokenKind::Data("World!".as_bytes().to_vec()),
                    line_column: (1, 19),
                    position: (18, 26)
                },
                Token {
                    kind: TokenKind::Semicolon,
                    line_column: (1, 27),
                    position: (26, 27)
                },
            ]
        );
    }

    #[test]
    fn use_parser() {
        let code = "a = 23;b = 1;";
        let expected = ["a = 23;", "b = 1;"];
        for (i, s) in super::parse(code).enumerate() {
            let stmt = s.unwrap();
            //assert!(matches!(stmt.kind(), Assign(..)));
            assert_eq!(&code[stmt.range()], expected[i]);
        }
    }
}
