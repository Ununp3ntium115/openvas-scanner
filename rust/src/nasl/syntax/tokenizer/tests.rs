use std::path::Path;

use codespan_reporting::files::SimpleFiles;

use crate::nasl::{error::emit_errors, syntax::utils::read_single_files};

use super::*;

fn tokenize(
    file_name: &str,
    code: &str,
) -> (
    SimpleFiles<String, String>,
    usize,
    Result<Vec<Token>, Vec<TokenizerError>>,
) {
    let (files, file_id) = read_single_files(Path::new(file_name), code);
    let tokens = Tokenizer::tokenize(code);
    (files, file_id, tokens)
}

fn tokenize_ok(file_name: &str, code: &str) -> Vec<Token> {
    let (files, file_id, results) = tokenize(file_name, code);
    match results {
        Ok(results) => results,
        Err(errors) => {
            emit_errors(&files, file_id, errors.into_iter());
            panic!()
        }
    }
}

fn tokenize_err(file_name: &str, code: &str) -> Vec<TokenizerError> {
    let (_, _, results) = tokenize(file_name, code);
    match results {
        Ok(_) => {
            panic!("Properly tokenized code that should result in error.")
        }
        Err(errors) => errors,
    }
}

macro_rules! test_ok {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(tokenize_ok(stringify!($name), $code));
        }
    };
}

macro_rules! test_err {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(tokenize_err(stringify!($name), $code));
        }
    };
}

test_ok!(skip_whitespace, "     (       ");

test_ok!(one_char_tokens, "( ) [ ] { } . - + % ; / * : ~ & | ^");
test_ok!(
    two_char_tokens,
    "& && | || ! != !~ = == =~ > >> >= >< < << <= - -- + += ++ / /= * ** *="
);
test_ok!(three_char_tokens, ">>> >>= >!< <<=");

test_ok!(four_symbol_tokens, ">>>=");

test_ok!(string, "\"I am a closed string \"");
test_ok!(unclosed_string, "\"hello I am a unclosed string\\");

test_ok!(data_string, r#"'Hello \\\'you\\\'!'"#);
test_ok!(unclosed_data_string, "'Hello \\'you\\'!\\'");

test_ok!(simple_number, "1");
test_ok!(
    numbers,
    "0 0b01 1234567890 012345670 0x1234567890ABCDEF 0b02"
);
test_err!(invalid_numbers, "0x 0b 0b2");
test_err!(invalid_hex_characters, "0x123h");

test_ok!(single_line_comments, "# this is a comment\n;");

test_ok!(identifier, "help_lo _hello _h4llo 4_h4llo");

test_ok!(keywords, "for foreach if else while repeat until local_var global_var NULL return include exit break continue");

test_ok!(string_quoting, r"'webapps\\\\appliance\\\\'");

test_ok!(data_escape_quoting, r#"'version=\"1.0\"'"#);

test_ok!(simplified_ipv4_address, "10.187.76.12");

test_ok!(repeat_x_times, "x() x 10;");
