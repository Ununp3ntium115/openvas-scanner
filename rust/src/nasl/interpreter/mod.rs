mod error;

mod assign;
mod call;
mod declare;
mod forking_interpreter;
mod include;
mod loop_extension;

#[cfg(test)]
mod tests;

use std::collections::{HashMap, VecDeque};

use crate::nasl::{
    Code, Context, Register,
    error::Span,
    prelude::NaslValue,
    syntax::{
        Ident,
        grammar::{
            Array, ArrayAccess, Ast, Atom, Binary, BinaryOperator, Block, Exit, Expr, If, Include,
            Return, Statement, Unary, UnaryPrefixOperator,
        },
    },
};

use error::IncludeSyntaxError;
pub use error::{FunctionCallError, InterpretError, InterpretErrorKind};
pub use forking_interpreter::ForkingInterpreter;

pub type Result<T = NaslValue, E = InterpretError> = std::result::Result<T, E>;

#[derive(PartialEq, Eq)]
enum InterpreterState {
    Running,
    Finished,
}

impl InterpreterState {
    fn is_finished(&self) -> bool {
        matches!(self, Self::Finished)
    }
}

/// Represents the result of a function call (`value`) along with
/// the `Span` pointing to the identifier of the function that
/// resulted in this value originally.
#[derive(Clone)]
pub struct FunctionCallData {
    value: NaslValue,
    span: Span,
}

/// This type contains the necessary data to reproduce the execution
/// flow in the case of interpreter forks.
///
/// Its two variants are
///
/// 1. `Collecting`: This variant is used by any interpreter which is
///    currently executing normally.  In this mode, the result of any
///    function call performed by the interpreter within a single
///    top-level statement will be stored along with the `Span` at which
///    the function call took place (as a safeguard). The variant also
///    stores the original `Register` and `Ast` in order to be able to "rewind"
///    into the exact state before the statement that caused the fork.
/// 2. `Restoring`: This variant is used by all interpreters which were just created due to a fork.
///    In this mode, the interpreter will not perform normal function
///    calls, and will instead return the first value in the `data` field
///    in place of the function call. This is done until the `data` field is
///    exhausted. At that point, execution proceeds normally.
#[derive(Clone)]
pub enum ForkReentryData {
    Collecting {
        data: Vec<FunctionCallData>,
        register: Register,
        // TODO: make this a cursor or something similar,
        // since cloning the entire AST seems wasteful.
        ast: Ast,
    },
    Restoring {
        data: VecDeque<FunctionCallData>,
    },
}

impl ForkReentryData {
    fn drain(&mut self) -> VecDeque<FunctionCallData> {
        match self {
            Self::Collecting { data, .. } => data.drain(..).collect(),
            _ => unreachable!(),
        }
    }

    fn register(&self) -> &Register {
        match self {
            Self::Collecting { register, .. } => register,
            _ => unreachable!(),
        }
    }

    fn ast(&self) -> &Ast {
        match self {
            Self::Collecting { ast, .. } => ast,
            _ => unreachable!(),
        }
    }

    fn contains_fork(&self) -> bool {
        match self {
            Self::Collecting { data, .. } => data
                .iter()
                .any(|value| matches!(value.value, NaslValue::Fork(_))),
            _ => false,
        }
    }

    /// If in `Collecting` mode, remember the given result. Otherwise
    /// do nothing.
    pub(crate) fn try_collect(&mut self, value: NaslValue, span: &Span) {
        match self {
            ForkReentryData::Collecting { data, .. } => data.push(FunctionCallData {
                value,
                span: span.clone(),
            }),
            ForkReentryData::Restoring { data: _ } => {}
        }
    }

    /// If in `Restoring` mode, remove and return the first stored
    /// result from the queue. Otherwise do nothing.
    pub(crate) fn try_restore(&mut self, span: &Span) -> Result<Option<NaslValue>, InterpretError> {
        match self {
            Self::Restoring { data } => {
                if let Some(data) = data.pop_front() {
                    if *span != data.span {
                        return Err(InterpretError::new(InterpretErrorKind::InvalidFork, None));
                    }
                    Ok(Some(data.value))
                } else {
                    Ok(None)
                }
            }
            _ => Ok(None),
        }
    }

    fn create_forks(&mut self) -> Vec<Self> {
        let mut data = vec![self.drain()];
        loop {
            let (changed, new_data) = expand_first_fork(data);
            data = new_data;
            if !changed {
                break;
            }
        }
        data.into_iter()
            .map(|data| Self::Restoring { data })
            .collect()
    }

    fn new() -> Self {
        Self::Restoring {
            data: VecDeque::new(),
        }
    }

    fn collecting(register: Register, ast: Ast) -> Self {
        Self::Collecting {
            data: vec![],
            register,
            ast,
        }
    }

    fn is_empty_or_collecting(&self) -> bool {
        match self {
            ForkReentryData::Collecting { .. } => true,
            ForkReentryData::Restoring { data } => data.is_empty(),
        }
    }
}

/// Expand the first occurrence of `NaslValue::Fork(...)` in the list of collected function
/// calls (i.e. any called function wants to fork), by returning one list per fork value.
///
/// For example (in pseudo-code):
/// expand_first_fork([[1, Fork(2, 3)]]) = [[1, 2], [1, 3]]
/// The boolean return value is true if any expansion took place or false if
/// no expansion took place (that is, if there was no `NaslValue::Fork` in the data).
fn expand_first_fork(
    data: Vec<VecDeque<FunctionCallData>>,
) -> (bool, Vec<VecDeque<FunctionCallData>>) {
    let first_fork = data[0]
        .iter()
        .enumerate()
        .filter_map(|(index, data)| {
            let FunctionCallData { value, span } = data;
            if let NaslValue::Fork(vals) = value {
                Some((index, vals.clone(), span.clone()))
            } else {
                None
            }
        })
        .next();
    if first_fork.is_none() {
        return (false, data);
    }
    let first_fork = first_fork.unwrap();
    let data = data
        .into_iter()
        .flat_map(|d| expand_fork_at(d, first_fork.0, first_fork.1.clone(), first_fork.2.clone()))
        .collect();
    (true, data)
}

fn expand_fork_at(
    data: VecDeque<FunctionCallData>,
    index: usize,
    vals: Vec<NaslValue>,
    span: Span,
) -> Vec<VecDeque<FunctionCallData>> {
    vals.iter()
        .map(|val| {
            let mut data = data.clone();
            data[index] = FunctionCallData {
                value: val.clone(),
                span: span.clone(),
            };
            data
        })
        .collect()
}

pub struct Interpreter<'ctx> {
    pub(super) register: Register,
    pub(super) context: &'ctx Context<'ctx>,
    pub(super) fork_reentry_data: ForkReentryData,
    ast: Ast,
    state: InterpreterState,
}

impl<'ctx> Interpreter<'ctx> {
    /// Creates a new Interpreter
    pub fn new(register: Register, ast: Ast, context: &'ctx Context) -> Self {
        Interpreter {
            register,
            ast,
            context,
            fork_reentry_data: ForkReentryData::new(),
            state: InterpreterState::Running,
        }
    }

    pub async fn execute_all(&mut self) -> Result<(), InterpretError> {
        while let Some(result) = self.execute_next_statement().await {
            result?;
        }
        Ok(())
    }

    pub async fn execute_next_statement(&mut self) -> Option<Result<NaslValue, InterpretError>> {
        self.initialize_fork_data();
        match self.ast.next_stmt() {
            Some(stmt) => {
                let result = self.resolve(&stmt).await;
                if matches!(result, Ok(NaslValue::Exit(_))) {
                    self.state = InterpreterState::Finished;
                }
                Some(result)
            }
            None => {
                self.state = InterpreterState::Finished;
                None
            }
        }
    }

    pub(super) async fn resolve(
        &mut self,
        statement: &Statement,
    ) -> Result<NaslValue, InterpretError> {
        use Statement::*;
        match statement {
            NoOp => Ok(NaslValue::Null),
            Continue => Ok(NaslValue::Continue),
            Break => Ok(NaslValue::Break),
            ExprStmt(expr) => self.resolve_expr(expr).await,
            Block(block) => self.resolve_block(block).await,
            VarScopeDecl(var_scope_decl) => self.resolve_var_scope_decl(var_scope_decl),
            FnDecl(fn_decl) => self.resolve_fn_decl(fn_decl),
            If(if_) => self.resolve_if(if_).await,
            While(while_) => Box::pin(self.resolve_while(while_)).await,
            Repeat(repeat) => Box::pin(self.resolve_repeat(repeat)).await,
            Foreach(foreach) => Box::pin(self.resolve_foreach(foreach)).await,
            For(for_) => Box::pin(self.resolve_for(for_)).await,
            Exit(exit) => self.resolve_exit(exit).await,
            Include(include_) => self.resolve_include(include_).await,
            Return(return_) => self.resolve_return(return_).await,
        }
        .map_err(|e: InterpretError| {
            if e.origin.is_none() {
                InterpretError::from_statement(statement, e.kind)
            } else {
                e
            }
        })
    }

    pub(super) async fn resolve_expr(&mut self, expr: &Expr) -> Result {
        match expr {
            Expr::Atom(atom) => Box::pin(self.resolve_atom(atom)).await,
            Expr::Binary(binary) => Box::pin(self.resolve_binary(binary)).await,
            Expr::Unary(unary) => Box::pin(self.resolve_unary(unary)).await,
            Expr::Assignment(assignment) => Box::pin(self.resolve_assignment(assignment)).await,
        }
    }

    pub async fn collect_exprs(
        &mut self,
        exprs: impl Iterator<Item = &Expr>,
    ) -> Result<Vec<NaslValue>> {
        let mut vals = vec![];
        for array_access in exprs {
            vals.push(self.resolve_expr(array_access).await?);
        }
        Ok(vals)
    }

    async fn resolve_atom(&mut self, atom: &Atom) -> Result {
        match atom {
            Atom::Literal(literal) => Ok(literal.into()),
            Atom::FnCall(call) => self.resolve_fn_call(call).await,
            Atom::Ident(ident) => Ok(self.resolve_var(ident)?.clone()),
            Atom::Array(array) => self.resolve_array(array).await,
            Atom::ArrayAccess(array_access) => self.resolve_array_access(array_access).await,
            Atom::Increment(increment) => self.resolve_increment(increment).await,
        }
    }

    fn resolve_var(&self, ident: &Ident) -> Result {
        let var = self.register.get(&ident.to_str());
        if let Some(var) = var {
            Ok(self.register.get_val(var).as_value()?.clone())
        } else {
            Ok(NaslValue::Null)
        }
    }

    async fn resolve_array_access(&mut self, array_access: &ArrayAccess) -> Result {
        let lhs = self.resolve_expr(&array_access.lhs_expr).await?;
        let index = self.resolve_expr(&array_access.index_expr).await?;
        lhs.index(index).map(|val| val.clone())
    }

    async fn resolve_unary(&mut self, unary: &Unary) -> Result {
        let rhs = self.resolve_expr(&unary.rhs).await?;
        match unary.op {
            UnaryPrefixOperator::Plus => Ok(rhs),
            UnaryPrefixOperator::Minus => rhs.neg(),
            UnaryPrefixOperator::Bang => rhs.not(),
            UnaryPrefixOperator::Tilde => rhs.bitwise_not(),
        }
    }

    async fn resolve_binary(&mut self, binary: &Binary) -> Result {
        use BinaryOperator::*;
        let lhs = self.resolve_expr(&binary.lhs).await?;
        // Short circuit
        if binary.op == AmpersandAmpersand && !lhs.as_boolean()? {
            return Ok(NaslValue::Boolean(false));
        }
        if binary.op == PipePipe && lhs.as_boolean()? {
            return Ok(NaslValue::Boolean(true));
        }
        let rhs = self.resolve_expr(&binary.rhs).await?;
        match binary.op {
            Plus => Ok(lhs.add(rhs)),
            Minus => Ok(lhs.sub(rhs)),
            Star => lhs.mul(rhs),
            Slash => lhs.div(rhs),
            Percent => lhs.rem(rhs),
            StarStar => lhs.pow(rhs),
            LessLess => lhs.shl(rhs),
            GreaterGreater => lhs.shr(rhs),
            GreaterGreaterGreater => lhs.shr_unsigned(rhs),
            Less => lhs.lt(rhs),
            LessEqual => lhs.le(rhs),
            Greater => lhs.gt(rhs),
            GreaterEqual => lhs.ge(rhs),
            Ampersand => lhs.bitand(rhs),
            Pipe => lhs.bitor(rhs),
            Caret => lhs.bitxor(rhs),
            AmpersandAmpersand => lhs.and(rhs),
            PipePipe => lhs.or(rhs),
            EqualTilde => lhs.match_regex(rhs),
            BangTilde => lhs.match_regex(rhs)?.not(),
            GreaterLess => lhs.match_string(rhs),
            GreaterBangLess => lhs.match_string(rhs)?.not(),
            EqualEqual => Ok(NaslValue::Boolean(lhs == rhs)),
            BangEqual => Ok(NaslValue::Boolean(lhs != rhs)),
        }
    }

    async fn resolve_array(&mut self, array: &Array) -> Result {
        Ok(NaslValue::Array(
            self.collect_exprs(array.items.items.iter()).await?,
        ))
    }

    async fn resolve_exit(&mut self, exit: &Exit) -> Result {
        let rc = Box::pin(self.resolve_expr(&exit.expr)).await?;
        match rc {
            NaslValue::Number(rc) => Ok(NaslValue::Exit(rc)),
            _ => Err(InterpretErrorKind::NonNumericExitCode(rc).into()),
        }
    }

    async fn resolve_return(&mut self, return_: &Return) -> Result {
        let rc = if let Some(ref expr) = return_.expr {
            self.resolve_expr(expr).await?
        } else {
            NaslValue::Null
        };
        Ok(NaslValue::Return(Box::new(rc)))
    }

    async fn resolve_if(
        &mut self,
        If {
            if_branches,
            else_branch,
        }: &If,
    ) -> Result<NaslValue, InterpretError> {
        for (condition, block) in if_branches.iter() {
            let val = self.resolve_expr(condition).await?;
            if bool::from(val) {
                return self.resolve_block(block).await;
            }
        }
        if let Some(block) = else_branch {
            return self.resolve_block(block).await;
        }
        Ok(NaslValue::Null)
    }

    pub(crate) async fn resolve_block(&mut self, block: &Block<Statement>) -> Result {
        self.register.create_child(HashMap::default());
        for stmt in block.items.iter() {
            match Box::pin(self.resolve(stmt)).await {
                Ok(x) => {
                    if matches!(
                        x,
                        NaslValue::Exit(_)
                            | NaslValue::Return(_)
                            | NaslValue::Break
                            | NaslValue::Continue
                    ) {
                        self.register.drop_last();
                        return Ok(x);
                    }
                }
                Err(e) => return Err(e),
            }
        }
        self.register.drop_last();
        // currently blocks return null
        Ok(NaslValue::Null)
    }

    async fn resolve_include(&mut self, include: &Include) -> Result {
        let loader = self.context.loader();
        let code = Code::load(loader, &include.path)?.parse();
        let file = code.file().clone();
        let ast = code.result().map_err(|errs| {
            InterpretErrorKind::IncludeSyntaxError(IncludeSyntaxError { file, errs })
        })?;
        let mut inter = Interpreter::new(self.register.clone(), ast, self.context);
        Box::pin(inter.execute_all()).await?;
        self.register = inter.register;
        Ok(NaslValue::Null)
    }

    pub(crate) fn make_forks(mut self) -> Vec<Interpreter<'ctx>> {
        let forks = self.fork_reentry_data.create_forks();
        let register = self.fork_reentry_data.register();
        let ast = self.fork_reentry_data.ast().clone();
        forks
            .into_iter()
            .map(|fork| self.make_fork(fork, register, &ast))
            .collect()
    }

    fn make_fork(
        &self,
        fork_reentry_data: ForkReentryData,
        register: &Register,
        ast: &Ast,
    ) -> Interpreter<'ctx> {
        Self {
            register: register.clone(),
            ast: ast.clone(),
            context: self.context,
            fork_reentry_data,
            state: InterpreterState::Running,
        }
    }

    pub(crate) fn wants_to_fork(&self) -> bool {
        self.fork_reentry_data.contains_fork()
    }

    pub(crate) fn initialize_fork_data(&mut self) {
        if self.fork_reentry_data.is_empty_or_collecting() {
            self.fork_reentry_data =
                ForkReentryData::collecting(self.register.clone(), self.ast.clone());
        }
    }

    pub(crate) fn is_finished(&self) -> bool {
        self.state.is_finished()
    }

    pub fn register(&self) -> &Register {
        &self.register
    }
}
