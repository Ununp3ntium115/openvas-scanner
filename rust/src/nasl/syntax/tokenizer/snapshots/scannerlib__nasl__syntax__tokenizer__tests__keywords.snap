---
source: src/nasl/syntax/tokenizer/tests.rs
expression: "tokenize_ok(stringify! (keywords),\n\"for foreach if else while repeat until local_var global_var return include exit break continue\")"
snapshot_kind: text
---
[
    Token {
        kind: Keyword(
            For,
        ),
        position: (
            0,
            3,
        ),
    },
    Token {
        kind: Keyword(
            ForEach,
        ),
        position: (
            4,
            11,
        ),
    },
    Token {
        kind: Keyword(
            If,
        ),
        position: (
            12,
            14,
        ),
    },
    Token {
        kind: Keyword(
            Else,
        ),
        position: (
            15,
            19,
        ),
    },
    Token {
        kind: Keyword(
            While,
        ),
        position: (
            20,
            25,
        ),
    },
    Token {
        kind: Keyword(
            Repeat,
        ),
        position: (
            26,
            32,
        ),
    },
    Token {
        kind: Keyword(
            Until,
        ),
        position: (
            33,
            38,
        ),
    },
    Token {
        kind: Keyword(
            LocalVar,
        ),
        position: (
            39,
            48,
        ),
    },
    Token {
        kind: Keyword(
            GlobalVar,
        ),
        position: (
            49,
            59,
        ),
    },
    Token {
        kind: Keyword(
            Return,
        ),
        position: (
            60,
            66,
        ),
    },
    Token {
        kind: Keyword(
            Include,
        ),
        position: (
            67,
            74,
        ),
    },
    Token {
        kind: Keyword(
            Exit,
        ),
        position: (
            75,
            79,
        ),
    },
    Token {
        kind: Keyword(
            Break,
        ),
        position: (
            80,
            85,
        ),
    },
    Token {
        kind: Keyword(
            Continue,
        ),
        position: (
            86,
            94,
        ),
    },
]
